great_expectations/datasource/new_datasource.py

```diff
@@ -261,6 +261,45 @@ def get_available_data_asset_names(
 
         return available_data_asset_names
 
+    def get_available_data_asset_names_and_types(
+        self, data_connector_names: Optional[Union[list, str]] = None
+    ) -> Dict[str, List[Tuple[str, str]]]:
+        """
+        Returns a dictionary of data_asset_names that the specified data
+        connector can provide. Note that some data_connectors may not be
+        capable of describing specific named data assets, and some (such as
+        inferred_asset_data_connector) require the user to configure
+        data asset names.
+
+        Returns:
+            dictionary consisting of sets of data assets available for the specified data connectors:
+            For instance, in a SQL Database the data asset name corresponds to the table or
+            view name, and the data asset type is either 'table' or 'view'.
+            ::
+
+                {
+                  data_connector_name: {
+                    names: [ (data_asset_name_1, data_asset_1_type), (data_asset_name_2, data_asset_2_type) ... ]
+                  }
+                  ...
+                }
+        """
+        # NOTE: Josh 20211001 This feature is only implemented for the InferredAssetSqlDataConnector
+
+        available_data_asset_names_and_types: dict = {}
+        if data_connector_names is None:
+            data_connector_names = self.data_connectors.keys()
+        elif isinstance(data_connector_names, str):
+            data_connector_names = [data_connector_names]
```

Comment:
Should we initialize `data_connector_names: List[str] = []` before these conditionals? If neither condition works out, `data_connector_names` will be undefined right?

Reply:
Thanks, @cdkini! That isn't an issue because `data_connector_names` is a kwarg with three valid types: a list, a string, or None. Those conditionals are there to coerce it into a list regardless of its initial type.